wolfcrypt/src/port/Espressif/esp32_sha.c

```diff
@@ -0,0 +1,399 @@
+
+        esp_dport_access_read_buffer((word32*)(hash), SHA_TEXT_BASE,
+                                              esp_sha_digest_size(sha_type)/4);
+
+#if defined(WOLFSSL_SHA512) || defined(WOLFSSL_SHA384)
+        if(sha_type==SHA2_384||sha_type==SHA2_512) {
+            word32  i;
+            word32* pwrd1 = (word32*)(hash);
+            /* swap value */
+            for(i = 0; i <WC_SHA512_DIGEST_SIZE/4; i+=2 ) {
+                pwrd1[i]  ^= pwrd1[i+1];
+                pwrd1[i+1]^= pwrd1[i];
+                pwrd1[i]  ^= pwrd1[i+1];
+            }
+        }
+#endif
+    }
+
+    ESP_LOGV(TAG, "leave esp_digest_state");
+}
+
+#ifndef NO_SHA
+/*
+* sha1 process
+*/
+int esp_sha_process(struct wc_Sha* sha)
+{
+    ESP_LOGV(TAG, "enter esp_sha_process");
+
+    int ret = 0;
+    word32 SHA_START_REG = SHA_1_START_REG + SHA1 * 0x10;
+
+    esp_process_block(&sha->ctx, SHA_START_REG, sha->buffer,
+                                        WC_SHA_BLOCK_SIZE);
+
+    ESP_LOGV(TAG, "leave esp_sha_process");
+    return ret;
+}
+/*
+* retrieve sha1 digest
+*/
+int esp_sha_digest_process(struct wc_Sha* sha, byte blockproc)
+{
+    ESP_LOGV(TAG, "enter esp_sha_digest_process");
+    int ret = 0;
+
+    if(blockproc) {
+        word32 SHA_START_REG = SHA_1_START_REG + SHA1 * 0x10;
+
+        esp_process_block(&sha->ctx, SHA_START_REG, sha->buffer,
+                                            WC_SHA_BLOCK_SIZE);
+    }
+
+    esp_digest_state(&sha->ctx, (byte*)sha->digest, SHA1);
+
+    ESP_LOGV(TAG, "leave esp_sha_digest_process");
+
+    return ret;
+}
+#endif /* NO_SHA */
+
+
+#ifndef NO_WOLFSSL_SHA256
+/*
+* sha256 process
+*/
+int esp_sha256_process(struct wc_Sha256* sha)
+{
+    ESP_LOGV(TAG, "enter esp_sha256_process");
+    int ret = 0;
+    word32 SHA_START_REG = SHA_1_START_REG + SHA2_256 * 0x10;
+
+    esp_process_block(&sha->ctx, SHA_START_REG, sha->buffer,
+                                            WC_SHA_BLOCK_SIZE);
+
+    ESP_LOGV(TAG, "leave esp_sha256_process");
+
+    return ret;
+}
+/*
+* retrieve sha256 digest
+*/
+int esp_sha256_digest_process(struct wc_Sha256* sha, byte blockproc)
+{
+    ESP_LOGV(TAG, "enter esp_sha256_digest_process");
+    int ret = 0;
+
+    if(blockproc) {
+        word32 SHA_START_REG = SHA_1_START_REG + SHA2_256 * 0x10;
+
+        esp_process_block(&sha->ctx, SHA_START_REG, sha->buffer,
+                                           WC_SHA_BLOCK_SIZE);
+    }
+
+    esp_digest_state(&sha->ctx, (byte*)sha->digest, SHA2_256);
+
+    ESP_LOGV(TAG, "leave esp_sha256_digest_process");
+    return ret;
+}
+#endif /* NO_WOLFSSL_SHA256 */
+
+#if defined(WOLFSSL_SHA512) || defined(WOLFSSL_SHA384)
+/*
+* sha512 proess. this is used for sha384 too.
+*/
+void esp_sha512_block(struct wc_Sha512* sha, const word32* data, byte isfinal)
+{
+    ESP_LOGV(TAG, "enter esp_sha512_block");
+
+    enum SHA_TYPE sha_type = sha->ctx.sha_type;
+    word32 SHA_START_REG = SHA_1_START_REG;
+
+    SHA_START_REG += sha_type * 0x10;
```

Comment:
Can you please use binary operations here. Or and left shift 4 here so the code is easier to understand. Should be `SHA_START_REG |= (sha_type << 4);`